pkg/terraform/diagnose.go

```diff
@@ -0,0 +1,76 @@
+package terraform
+
+import (
+	"regexp"
+
+	"github.com/pkg/errors"
+
+	"github.com/openshift/installer/pkg/diagnostics"
+)
+
+// Diagnose accepts an error from terraform runs and tries to diagnose the
+// underlying cause.
+func Diagnose(message string) error {
+	for _, cand := range conditions {
+		if cand.match.MatchString(message) {
+			return &diagnostics.Err{
+				Source:  "InfrastructureProvider",
+				Reason:  cand.reason,
+				Message: cand.message,
+			}
+		}
+	}
+
+	return errors.New("failed to complete the change")
+}
+
+type condition struct {
+	match *regexp.Regexp
+
+	reason  string
+	message string
+}
+
+// conditions is a list matches for the error string from terraform.
+// specific on the top, generic matches on the bottom.
+var conditions = []condition{{
+	match: regexp.MustCompile(`Error: Error creating Blob .*: Error copy/waiting`),
+
+	reason:  "Timeout",
+	message: `Copying the VHD to user environment was too slow, and timeout was reached for the success.`,
+}, {
+	match: regexp.MustCompile(`Error: Error Creating/Updating Subnet .*: network.SubnetsClient#CreateOrUpdate: .* Code="AnotherOperationInProgress" Message="Another operation on this or dependent resource is in progress`),
+
+	reason:  "AzureMultiOperationFailure",
```

Comment:
Can we structure this so Azure conditions can be defined in a `pkg/terraform/azure` subpackage, so per-platform sub-teams can be assigned as owners for their platform's conditions?  There might be some generic conditions that apply to all platforms, but I'd expect most of the time, you'd want a whole bunch of platform-specific conditions from the platform the installer is using, plus maybe a handful of generic conditions.

Reply:
for i'm keeping this list in same file and we can split in into platforms in future work.